// @tatumio/shared-testing-evm-based/src/lib/evm-based.native.test-factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nativeTestFactory = void 0;
const tslib_1 = require("tslib");
const shared_testing_common_1 = require("@tatumio/shared-testing-common");
exports.nativeTestFactory = {
    prepare: {
        transferSignedTransaction: (sdk, accounts) => {
            it('valid from signatureId', () => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
                const nonce = 3252345722143;
                const result = yield sdk.prepare.transferSignedTransaction({
                    to: '0x811DfbFF13ADFBC3Cf653dCc373C03616D3471c9',
                    amount: '1',
                    signatureId: 'cac88687-33ed-4ca1-b1fc-b02986a90975',
                    nonce,
                    fee: {
                        gasLimit: '53632',
                        gasPrice: '20',
                    },
                });
                const json = JSON.parse(result);
                expect(json.nonce).toBe(nonce);
                expect(json.gasPrice).toBe('20000000000');
                expect(json.value).toBe('1000000000000000000');
            }));
            it('valid from privateKey', () => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
                const nonce = 3252345722143;
                const result = yield sdk.prepare.transferSignedTransaction({
                    to: '0x811DfbFF13ADFBC3Cf653dCc373C03616D3471c9',
                    amount: '1',
                    fromPrivateKey: accounts[0].privateKey,
                    nonce,
                    fee: {
                        gasLimit: '53632',
                        gasPrice: '20',
                    },
                });
                (0, shared_testing_common_1.expectHexString)(result);
            }));
            it('invalid address', () => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
                try {
                    yield sdk.prepare.transferSignedTransaction({
                        to: 'someinvalidaddress',
                        amount: '1',
                        fromPrivateKey: accounts[0].privateKey,
                        nonce: 3252345722143,
                        fee: {
                            gasLimit: '53632',
                            gasPrice: '20',
                        },
                    });
                    fail();
                }
                catch (e) {
                    expect(e.message).toMatch('is invalid');
                }
            }));
        },
    },
};
