// @tatumio/shared-core/src/lib/models/Blockchain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BtcBasedBlockchains = exports.EvmBasedBlockchains = exports.Blockchain = void 0;
var Blockchain;
(function (Blockchain) {
    Blockchain["BNB"] = "BNB";
    Blockchain["BTC"] = "BTC";
    Blockchain["BCH"] = "BCH";
    Blockchain["LTC"] = "LTC";
    Blockchain["CELO"] = "CELO";
    Blockchain["SOL"] = "SOL";
    Blockchain["HARMONY"] = "HARMONY";
    Blockchain["ETH"] = "ETH";
    Blockchain["XRP"] = "XRP";
    Blockchain["XLM"] = "XLM";
    Blockchain["DOGE"] = "DOGE";
    Blockchain["VET"] = "VET";
    Blockchain["NEO"] = "NEO";
    Blockchain["BSC"] = "BSC";
    Blockchain["POLYGON"] = "POLYGON";
    Blockchain["FLOW"] = "FLOW";
    Blockchain["TRON"] = "TRON";
    Blockchain["EGLD"] = "EGLD";
    Blockchain["XDC"] = "XDC";
    Blockchain["ADA"] = "ADA";
    Blockchain["ALGO"] = "ALGO";
    Blockchain["KCS"] = "KCS";
    Blockchain["KLAY"] = "KLAY";
    Blockchain["TEZOS"] = "TEZOS";
    Blockchain["EON"] = "EON";
})(Blockchain = exports.Blockchain || (exports.Blockchain = {}));
exports.EvmBasedBlockchains = [
    Blockchain.ETH,
    Blockchain.CELO,
    Blockchain.BSC,
    Blockchain.HARMONY,
    Blockchain.POLYGON,
    Blockchain.KCS,
    Blockchain.EON,
];
// @TODO tmp solution
exports.BtcBasedBlockchains = [Blockchain.BTC, Blockchain.LTC, Blockchain.DOGE, Blockchain.BCH];
